src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
+            elementData[size] = value;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < ZERO || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of ArrayList length");
+        }
 
+        if (size == elementData.length) {
+            elementData = Arrays.copyOf(elementData, newSize(elementData.length));
+        }
+
+        if (index == size) {
+            elementData[size] = value;
+        } else {
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+            elementData[index] = value;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null && !list.isEmpty()) {
+
+            if (elementData.length < size + list.size()) {
+                elementData = Arrays.copyOf(elementData, size + list.size());
+            }
 
+            for (int i = ZERO; i < list.size(); i++) {
+                T element = list.get(i);
+                elementData[size++] = element;
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= ZERO && index < size) {
+            return elementData[index];
+        }
+        throw new ArrayListIndexOutOfBoundsException("Index is out of ArrayList length");
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < ZERO) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of ArrayList length");
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < ZERO || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of ArrayList length");
+        }
+        T removedElement = elementData[index];
+
+        System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+
+        elementData[--size] = null;
+
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = ZERO; i < size; i++) {
+            if ((element == null && elementData[i] == null)
+                    || (element != null && element.equals(elementData[i]))) {
+                T removedElement = elementData[i];
+                System.arraycopy(elementData, i + 1, elementData, i, size - i - 1);
+                elementData[--size] = null;
+
+                return removedElement;
+            }
+        }
+        throw new NoSuchElementException("Can't find the element in ArrayList");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == ZERO;
+    }
+
+    public int newSize(int size) {
```

Comment:
why public? IMO logic too simple to extract to a separate method, also not sure why you need + 1, how can size * RE_SIZE not be more than size?

Reply:
becouse of AI:
"The grow method should ensure that the new capacity is greater than the current size. Consider adding a check to ensure the new capacity is at least size + 1 to prevent potential issues when the list is full and needs to grow."